Tp 00043 recycle fun facts on info

annie951977 opened the pull request:
# Description
Works now for real

Fixes #43 

## Type of change

Please delete options that are not relevant.
- [X] New feature (non-breaking change which adds functionality)

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation (if required)
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works

annie951977 committed:
Merge conflicts

annie951977 committed:
InfoService

annie951977 committed:
Add DYKFacts.json and DYKInfo

annie951977 committed:
Merge branch 'develop' into TP-00043-recycle-fun-facts-on-info

annie951977 committed:
Remove react native midnight

annie951977 committed:
Weird stuff

annie951977 committed:
Moved info service

annie951977 committed:
Added another helper

annie951977 committed:
Get fact shortened function

annie951977 committed:
get fact shortened function

annie951977 committed:
fixed path error on DYKInfo

annie951977 committed:
Trying to fix promise rejection

annie951977 committed:
Merge branch 'develop' into TP-00043-recycle-fun-facts-on-info

annie951977 committed:
Update package-lock.json

annie951977 committed:
Merge branch 'develop' into TP-00043-recycle-fun-facts-on-info

annie951977 committed:
Trying to get the fact cycle to work

faye-tan commented on tpp-app/src/info/InfoService.js:
```diff
@@ -0,0 +1,69 @@
+     if (await AsyncStorage.getItem(FACT_CYCLE.FACT_CYCLE_NUM) == null) {
+        resolve(null);
+     } else {
+         value = await AsyncStorage.getItem(FACT_CYCLE.FACT_CYCLE_NUM).then((value) => {
```
not sure if this is your error but I don't think you should be using await and .then() together (it's basically doing an await twice or a .then() twice), just as try/except does the same job as .catch(). I would do it like this, try it and lmk if it fixes your error: 
```
export const GETFactCycle = async () => new Promise(async (resolve, reject) => {
    try {
     if (await AsyncStorage.getItem(FACT_CYCLE.FACT_CYCLE_NUM) == null) {
        resolve(null);
     } else {
         value = await AsyncStorage.getItem(FACT_CYCLE.FACT_CYCLE_NUM);
         console.log("Retrieved Fact Cycle Date and Number");
         resolve(value);
     }
    } catch (e) {
        console.log(`GETFactCycle error: ${JSON.stringify(e)}`)
        reject(null);
    }
});
```

faye-tan commented on tpp-app/src/info/Info.js:
```diff
@@ -29,15 +33,42 @@ const LearnMoreCard = () => {
+    if (factCycle) {
+        await POSTFactCycle() // POSTFact cycle iniaties fact cycle
+
+        factCycle = await GETFactCycle();
```
There are a lot of asynchronous calls being made here, which may conflict with each other as each of them retrieves and updates in AsyncStorage. They may resolve in an arbitrary order, such as line 52 may not finish before line 54 executes. 

The purpose of .then() is to chain async calls together which would be helpful in guaranteeing an async call finishes. 

```
POSTFactCycle()
  .then(() => { 
     await GETFactCycle();  
})
```

The .then() ensures that POSTFactCycle completes before calling GETFactCycle.

helen-li commented on tpp-app/src/info/Info.js:
```diff
@@ -29,15 +33,42 @@ const LearnMoreCard = () => {
 
+const getShortenFact = async () => {
+    var factCycle = await GETFactCycle();
+    if (factCycle) {
```
i may be tripping but i think what this code does right now is that if fact cycle exists, we enter this if block which then initiates fact cycle again with POSTFactCycle? is this the intended behavior or did you mean to do `if (!factCycle)` annie?

annie951977 replied:
I'm tripping, i meant `!factCycle`

helen-li commented on tpp-app/src/info/InfoService.js:
```diff
@@ -0,0 +1,69 @@
+export const POSTFactCycle = async () => new Promise(async (resolve, reject) => {
+    try {
+        if (await GETFactCycle() == null) { // if there is currently no fact/date array stored
+            const value = [];
```
something nitpicky lol no need to change this at all it's just my OCD coming out but is there a reason we cannot just do `const value = [getFullCurrentDateString(), "1"]` here

annie951977 replied:
me at 7 am for a second thought we couldnt lmao

helen-li commented on tpp-app/src/services/utils/helpers.js:
```diff
@@ -197,4 +185,20 @@ export const GETStoredYears = async () => {
+* @returns a string representing the current date
+*/
+
+export function getFullCurrentDateString() {
```
i believe `getDateString(new Date())` will do what this helper does? also not sure if this is intended, but i don't think this helper takes into account that getMonth is 0-indexed in javascript (i.e. January = 0, February = 1, etc.)

annie951977 replied:
rip lmao

annie951977 replied:
WOuld it be fine if I just wrote a comment that the dates are 0-indexed?

annie951977 committed:
Adding .then()

annie951977 committed:
CR Comment

annie951977 committed:
Brain fart lmao

annie951977 committed:
Wrote async

annie951977 committed:
I changed this but still catches

annie951977 committed:
Use effect

annie951977 committed:
This is working now huzzah

annie951977 committed:
I'm actually a dumbass jfc

Changed to use toString() right. Added a safeguard in case there's a problem calling previousFactNum

annie951977 committed:
Console logs

annie951977 committed:
Retrieve factCycle on info page

annie951977 committed:
Fixed bug

faye-tan commented on tpp-app/src/info/Info.js:
```diff
@@ -82,11 +89,44 @@ const cardData = [
+    const [factCycleArray, setFactCycleArray] = useState([]);
+    useEffect(() => {
+        async function retrieveFactCycle() {
+            var factArray = await GETFactCycle()
```
can you change all ```var``` to ```let```? since vars are global scoped and we'd prefer locally scoped variables by using ```let```

faye-tan commented on tpp-app/src/info/Info.js:
```diff
@@ -82,11 +89,44 @@ const cardData = [
+            console.log(`This is setting factCycleArray on the InfoPage: ${factArray}`)
+
+            if (!factArray){
+                await POSTFactCycle().then(async () => {
```
don't need await here

faye-tan commented on tpp-app/src/info/InfoService.js:
```diff
@@ -0,0 +1,71 @@
+        } else {
+            // get the current date
+            newDate = getFullCurrentDateString()
+            if (newDate != GETFactCycle()[0]) { // update dateand fact only if the current date does not match the stored date
```
GETFactCycle() prob needs await in front of it

roylin2000 commented on tpp-app/src/info/Info.js:
```diff
@@ -82,11 +89,44 @@ const cardData = [
+    }, [])
+
+
+    let factWhole = factsJSON[factCycleArray[1]]
```
For me, the useEffect is not fast enough and factCycleArray is not defined at this time yet, so an error pops. I think you need to do this in the useEffect as well. I was reading this and it might be relevant. https://devtrium.com/posts/async-functions-useeffect

faye-tan commented on tpp-app/src/info/InfoService.js:
```diff
@@ -0,0 +1,71 @@
+
+export const POSTFactCycle = async () => new Promise(async (resolve, reject) => {
+    try {
+        if (await GETFactCycle() == null) { // if there is currently no fact/date array stored
```
better to use strict equality ===

annie951977 committed:
Keeps falling back on the random generator, looking into why it can't get the next num

annie951977 committed:
This is cycling now

annie951977 committed:
At least made sure that it's nonbreaking right now

faye-tan committed:
messy messy faye additions

faye-tan committed:
messy messy faye additions

annie951977 committed:
Info page keeps falling back on the default

annie951977 committed:
Last fact

annie951977 committed:
a little bit of cleanup

annie951977 committed:
Merge branch 'develop' into TP-00043-recycle-fun-facts-on-info

annie951977 committed:
Merge conflicts

annie951977 committed:
Merge conflict

roylin2000 reviewed (approved):
All good now!

helen-li reviewed (approved):
nicee looks good annie!

faye-tan commented on tpp-app/src/info/Info.js:
```diff
@@ -85,12 +91,69 @@ const cardData = [
+                            setFactCycleNum(rePostedFactArray[1]);
+                            factWhole = factsJSON[rePostedFactArray[1]]
+                            console.log(`This is fact on Info Page 112: ${factWhole}`)
+                            setRetrievedFact(factWhole.slice(0, 84));
```
can you slice by factWhole.length // 2? Since some are less than 84 chars

faye-tan reviewed (changes requested):
remove console.log statements and some small things and i think its good

annie951977 committed:
Small changes

annie951977 committed:
formatting

annie951977 committed:
Kenneth you're a maryland king
